modules/michaoBidAdapter.js

```diff
@@ -0,0 +1,352 @@
+    );
+  },
+
+  invalidPartnerError(value) {
+    logError(
+      `Michao Bid Adapter: Invalid partner ID. Expected number, got ${typeof value}. Value: ${value}`
+    );
+  },
+
+  invalidBadvError(value) {
+    logError(
+      `Michao Bid Adapter: Invalid badv. Expected array, got ${typeof value}`
+    );
+  },
+
+  invalidBcatError(value) {
+    logError(
+      `Michao Bid Adapter: Invalid bcat. Expected array, got ${typeof value}`
+    );
+  },
+
+  invalidRewardError(value) {
+    logError(
+      `Michao Bid Adapter: Invalid reward. Expected boolean, got ${typeof value}. Value: ${value}`
+    );
+  },
+
+  invalidVideoContext() {
+    logError('Michao Bid Adapter: Video context is not set');
+  },
+
+  invalidVideoPlayerSize() {
+    logError('Michao Bid Adapter: Video playerSize is not set or invalid');
+  },
+
+  invalidVideoMinDuration() {
+    logError('Michao Bid Adapter: Video minDuration is not set or invalid');
+  },
+
+  invalidVideoMaxDuration() {
+    logError('Michao Bid Adapter: Video maxDuration is not set or invalid');
+  },
+
+  invalidVideoMimes() {
+    logError('Michao Bid Adapter: Video mimes is not set or invalid');
+  },
+
+  invalidVideoProtocols() {
+    logError('Michao Bid Adapter: Video protocols is not set or invalid');
+  },
+};
+
+function buildRequest(bidRequests, bidderRequest) {
+  const openRTBBidRequest = converter.toORTB({
+    bidRequests: bidRequests,
+    bidderRequest,
+  });
+
+  return {
+    method: 'POST',
+    url: ENV.ENDPOINT,
+    data: openRTBBidRequest,
+    options: { contentType: 'application/json', withCredentials: true },
+  };
+}
+
+function generateGdprParams(gdprConsent) {
+  let gdprParams = '';
+
+  if (typeof gdprConsent === 'object') {
+    if (gdprConsent?.gdprApplies) {
+      gdprParams = `gdpr=${Number(gdprConsent.gdprApplies)}&gdpr_consent=${
+        gdprConsent.consentString || ''
+      }`;
+    }
+  }
+
+  return gdprParams;
+}
+
+function generateBillableUrl(bid) {
+  return replaceAuctionPrice(bid.burl, bid.originalCpm || bid.cpm);
+}
+
+const converter = ortbConverter({
+  request(buildRequest, imps, bidderRequest, context) {
+    const bidRequest = context.bidRequests[0];
+    const openRTBBidRequest = buildRequest(imps, bidderRequest, context);
+    openRTBBidRequest.cur = [ENV.DEFAULT_CURRENCY];
+    openRTBBidRequest.test = config.getConfig('debug') ? 1 : 0;
+    openRTBBidRequest.bcat = bidRequest.params?.bcat || [];
+    openRTBBidRequest.badv = bidRequest.params?.badv || [];
+    deepSetValue(
+      openRTBBidRequest,
+      'site.id',
+      bidRequest.params.site.toString()
+    );
+    if (bidRequest?.schain) {
+      deepSetValue(openRTBBidRequest, 'source.schain', bidRequest.schain);
+    }
+
+    if (bidRequest.params.partner) {
+      deepSetValue(
+        openRTBBidRequest,
+        'site.publisher.ext.partner',
+        bidRequest.params.partner.toString()
+      );
+    }
+
+    return openRTBBidRequest;
```

Comment:
Your doc says you only support site and user, but you seem to support device and other parts of ortb as well?